volatility3/framework/plugins/windows/pedump.py

```diff
@@ -0,0 +1,178 @@
+from volatility3.framework.configuration import requirements
+from volatility3.framework.symbols import intermed
+from volatility3.framework.symbols.windows.extensions import pe
+from volatility3.plugins.windows import pslist, modules
+
+vollog = logging.getLogger(__name__)
+
+
+class PEDump(interfaces.plugins.PluginInterface):
+    """Allows extracting PE Files from a specific address in a specific address space"""
+
+    _required_framework_version = (2, 0, 0)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        # Since we're calling the plugin, make sure we have the plugin's requirements
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.VersionRequirement(
+                name="pslist", component=pslist.PsList, version=(2, 0, 0)
+            ),
+            requirements.ListRequirement(
+                name="pid",
+                element_type=int,
+                description="Process IDs to include (all other processes are excluded)",
+                optional=True,
+            ),
+            requirements.IntRequirement(
+                name="base",
+                description="Base address to reconstruct a PE file",
+                optional=False,
+            ),
+            requirements.BooleanRequirement(
+                name="kernel_module",
+                description="Extract from kernel address space.",
+                default=False,
+                optional=True,
+            ),
+        ]
+
+    def _write_pe(self, pe_table_name, layer_name, proc_offset, pid):
+        try:
+            file_handle = self.open(
+                "PE.{:#x}.{:d}.{:#x}.dmp".format(
+                    proc_offset,
+                    pid,
+                    self.config["base"],
+                )
+            )
+
+            dos_header = self.context.object(
+                pe_table_name + constants.BANG + "_IMAGE_DOS_HEADER",
+                offset=self.config["base"],
+                layer_name=layer_name,
+            )
+
+            for offset, data in dos_header.reconstruct():
+                file_handle.seek(offset)
+                file_handle.write(data)
+        except (
+            IOError,
+            exceptions.VolatilityException,
+            OverflowError,
+            ValueError,
+        ) as excp:
+            vollog.debug(f"Unable to dump dll at offset {self.config['base']}: {excp}")
+            return None
+
+        file_handle.close()
+        return file_handle.preferred_filename
+
+    def _dump_kernel(self, _, pe_table_name, session_layers):
+        session_layer_name = modules.Modules.find_session_layer(
+            self.context, session_layers, self.config["base"]
```

Comment:
Ideally this would be passed `base` as a parameter.  Although this is private, threading it that way would make it more reusable by others (and since it is a swappable/reusable function, that seems like it would be useful in the future).

Reply:
done